### PR TITLE
[codex] Make local-dev capability policy declarative

### PR DESCRIPTION
**Pull request by serrrfirat**
## Summary
- Move Reborn local-dev built-in capability grants and provider trust effects into a bundled TOML policy.
- Parse and validate the host-owned policy in composition, then reuse it for visible capabilities, first-party trust, skill-management mounts, and approval lease terms.
- Add a declarative spawn-capability approval default so non-visible spawn approvals still receive lease terms without hardcoded Rust capability classes.

## Validation
- cargo fmt -p ironclaw_reborn_composition
- cargo test -p ironclaw_reborn_composition
- git diff --check

### CODE REVIEW

File: crates/ironclaw_reborn_composition/src/runtime/approval.rs [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
-use ironclaw_approvals::LeaseApproval;
-use ironclaw_host_api::{Action, CapabilityId, EffectKind, MountView, Principal};
+use ironclaw_host_api::MountView;
 use ironclaw_product_workflow::{
     ApprovalGateRecord, ApprovalInteractionRejectionKind, ApprovalLeaseTermsProvider,
     ProductWorkflowError,
 };
 
-use super::local_dev;
+use crate::local_dev_capability_policy::{
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/runtime/approval.rs]: **Medium** — Policy parsed from TOML on every lease_terms_for call.

local_dev_capability_policy() re-parses and re-validates the bundled TOML on every call to lease_terms_for. While the policy is compile-time constant (include_str!), the repeated toml::from_str + validation overhead on every approval action is unnecessary. The LocalDevApprovalLeaseTermsProvider struct already holds per-call-invariant mounts fields; the parsed policy could also be held there.

**Fix:** Parse once at construction time and store the Arc<LocalDevCapabilityPolicy> (or just a copy, since it's Clone) on LocalDevApprovalLeaseTermsProvider. Same fix applies to LocalDevLoopCapabilityPortFactory::create_capability_port if the factory is long-lived.

File: crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    }
+
+    fn network(&self) -> LocalDevNetworkProfile {
+        self.network
+    }
+}
+
+pub(crate) fn local_dev_capability_policy()
+-> Result<LocalDevCapabilityPolicy, LocalDevCapabilityPolicyError> {
+    parse_local_dev_capability_policy(LOCAL_DEV_CAPABILITY_POLICY_TOML)
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs]: **Medium** — Public parse function is not guarded by OnceLock — repeated static parse.

local_dev_capability_policy() is a module-level function that parses an include_str! constant on every call. Three call sites (approval.rs, local_dev.rs x2 via factory.rs) each call it independently with no shared cache. Since the result is deterministic for the lifetime of the binary, this is a natural OnceLock<Result<...>> candidate. Current call pattern is not wrong for correctness, but sets a precedent for more callers being added without noticing the cost.

**Fix:** Wrap the parse result in a OnceLock<LocalDevCapabilityPolicy> initialized on first call and return a reference or clone. Or ensure callers pre-parse at construction time (as noted above) so the function is never called in hot paths.

File: crates/ironclaw_reborn_composition/src/runtime/approval.rs (hunk 1 above)
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/runtime/approval.rs]: **Medium** — Policy parse error is silently swallowed — no log, wrong rejection kind.

When local_dev_capability_policy() returns an error (corrupt embedded TOML or validator failure), it is mapped to ApprovalInteractionRejectionKind::LeaseTermsUnavailable with no log line. If the TOML is malformed (e.g. after a bad merge), every approval silently fails with an opaque rejection and no observable trace. The same error is also swallowed in lease_approval_for's map_err.

**Fix:** Add a tracing::warn! or tracing::error! before the map_err so an operator has a chance to see what broke. Since this is a host-internal error (not user-input), it warrants error! level. Also consider keeping the policy as a pre-validated field on the provider struct to prevent this code path entirely.

File: crates/ironclaw_reborn_composition/src/runtime/local_dev.rs [hunk 3]
```diff
@@ -7,17 +7,12 @@ use chrono::Utc;
...
+    UserId,
 };
 use ironclaw_host_runtime::{
-    APPLY_PATCH_CAPABILITY_ID, CapabilitySurfacePolicy, ECHO_CAPABILITY_ID, GLOB_CAPABILITY_ID,
-    GREP_CAPABILITY_ID, HTTP_CAPABILITY_ID, HostRuntime, JSON_CAPABILITY_ID,
-    LIST_DIR_CAPABILITY_ID, READ_FILE_CAPABILITY_ID, SHELL_CAPABILITY_ID,
-    SKILL_INSTALL_CAPABILITY_ID, SKILL_LIST_CAPABILITY_ID, SKILL_REMOVE_CAPABILITY_ID, SurfaceKind,
-    TIME_CAPABILITY_ID, VisibleCapabilityRequest as HostVisibleCapabilityRequest,
-    WRITE_FILE_CAPABILITY_ID,
+    CapabilitySurfacePolicy, HostRuntime, SurfaceKind,
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/runtime/local_dev.rs:14]: **Medium** — Policy re-parsed on every capability port creation — factory is not the right parse site.

local_dev_capability_policy() is called inside create_capability_port which executes for every run context / turn. Since LocalDevLoopCapabilityPortFactory is constructed once per session/user and held in an Arc, parsing the static policy inside the per-turn create_capability_port call wastes CPU and, more importantly, any TOML parse failure becomes a per-turn error rather than a startup-time error.

**Fix:** Parse the policy in LocalDevLoopCapabilityPortFactory::new (or a factory builder) and store it as a field. Surface any parse error at construction time, consistent with local_dev_first_party_trust_policy() which already uses map_err at build time.

File: crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        &self,
+        action: LocalDevApprovalPolicyAction<'_>,
+        workspace_mounts: &MountView,
+        skill_mounts: &MountView,
+    ) -> Result<LeaseApproval, LocalDevCapabilityPolicyError> {
+        let constraints = match action {
+            LocalDevApprovalPolicyAction::Dispatch { capability } => {
+                self.grant_constraints_for(capability, workspace_mounts, skill_mounts)?
+            }
+            LocalDevApprovalPolicyAction::SpawnCapability { capability } => {
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:119]: **Low** — SpawnCapability approval ignores extra SpawnProcess injection when capability is in the grants list.

In lease_approval_for, for the SpawnCapability action variant: when the capability IS found in the grants list, constraint_terms is called with required_effect = Some(EffectKind::SpawnProcess) so SpawnProcess is forcibly added. When the capability is NOT in grants (fallback to approval_defaults.spawn_capability), constraint_terms is called with required_effect = None, so whatever the defaults declare is used as-is. This asymmetry is intentional per the comment, but the required_effect injection on the 'found' branch only appends SpawnProcess if it is not already present — the semantics differ subtly from the old code which explicitly appended. Verify the intent: if the policy already declares SpawnProcess in the grant's effects for a spawnable capability, the dedup is correct; if it does not and that is intentional, the dedup masks a misconfiguration.

**Fix:** Add a test that verifies a grant without SpawnProcess in its effects still receives it in the lease when called via SpawnCapability. Also add a test that a grant that already declares SpawnProcess does not get it duplicated.

File: crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+            max_invocations: Some(1),
+        })
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct LocalDevProviderPolicy {
+    pub(crate) id: PackageId,
+    pub(crate) manifest_path: String,
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:160]: **Low** — manifest_path field is a plain String instead of a domain newtype.

The LocalDevProviderPolicy struct holds manifest_path: String. Per .claude/rules/types.md, internal domain values should use newtypes rather than raw String to prevent misuse at layer boundaries. While this is an internal-only policy struct, the manifest_path value ends up passed as-is to AdminEntry::for_local_manifest (in factory.rs) without further validation beyond the is_empty() trim check in validate_policy.

**Fix:** Consider wrapping it in a VirtualPath or a validated ManifestPath newtype (if one exists), or at minimum enforce the leading-slash invariant in validate_policy rather than just checking for empty.

File: crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+        max_egress_bytes: None,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn bundled_local_dev_capability_policy_parses() {
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:382]: **Low** — bundled_local_dev_capability_policy_parses test does not validate approval_defaults.spawn_capability mounts/network fields.

The test at line 1880 checks that spawn_capability.effects contains SpawnProcess but does not assert the mounts or network field values for the approval defaults section. If someone changes the TOML defaults (e.g. sets network to local_dev_wildcard), the test would still pass. Given that the approval defaults govern what lease terms are issued for process spawn approvals, misconfiguring those fields is high-impact.

**Fix:** Add assertions for policy.approval_defaults.spawn_capability.mounts == LocalDevMountProfile::Workspace and .network == LocalDevNetworkProfile::Default to lock the defaults.

File: crates/ironclaw_reborn_composition/src/local_dev_capability_policy.toml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+network = "default"
+
+[[grants]]
+capability = "builtin.http"
+effects = ["dispatch_capability", "network"]
+mounts = "ambient"
+network = "local_dev_wildcard"
+
+[[grants]]
+capability = "builtin.shell"
```
Comment (henrypark133) [inline crates/ironclaw_reborn_composition/src/local_dev_capability_policy.toml:57]: **Low** — builtin.shell grant includes write_filesystem — broader than execution-only.

The shell grant declares [dispatch_capability, read_filesystem, write_filesystem, spawn_process, execute_code, network] with mounts=ambient. The ambient mounts profile means MountView::default() (no scoped mounts), so write_filesystem permission is granted but no mount constraint backs it. This matches the previous Rust implementation, so this is not a regression, but now that the policy is declarative, it is worth noting that the combination of write_filesystem + ambient + local_dev_wildcard network is the broadest local-dev grant. The comment in the old code ('The first-party shell handler still uses direct host process execution') justifying ambient was not carried forward into the TOML.

**Fix:** Add a comment in local_dev_capability_policy.toml above the builtin.shell entry explaining the ambient+write_filesystem intent (same as the removed Rust comment) so future maintainers have the context when reviewing. Low risk — this is local-dev only and matches the prior behavior.

**Review by henrypark133 — COMMENTED** [review]
## Code Review (sonnet single-agent)

**Intent:** Move the local-dev built-in capability grants and first-party trust effects out of hardcoded Rust into a declarative TOML policy (`local_dev_capability_policy.toml`); parse and validate the policy at composition time; reuse it across visible capabilities, grants, skill-management mounts, and approval lease terms. Also wire the `system.process_sandbox.run` capability through a new `spawn_capability` / `resume_spawn_capability` path on the `HostRuntime` trait with full input validation and test coverage.

**Stats:** 8 findings across 4 files

### Bugs / Correctness
- **Medium** Policy parse error silently swallowed — no log, wrong rejection kind (`crates/ironclaw_reborn_composition/src/runtime/approval.rs:36-45`, confidence 78) — `local_dev_capability_policy()` failures mapped to `LeaseTermsUnavailable` with no tracing output; a corrupt TOML would silently fail every approval with no observable trace.
- **Medium** Policy re-parsed on every capability port creation — factory is not the right parse site (`crates/ironclaw_reborn_composition/src/runtime/local_dev.rs:126`, confidence 72) — `local_dev_capability_policy()` called inside `create_capability_port` (per-turn) rather than at factory construction time; TOML failures become per-turn errors rather than startup-time errors.
- **Low** SpawnCapability approval: missing test for SpawnProcess injection on grants that lack it (`crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:1629-1651`, confidence 70) — `constraint_terms(..., Some(SpawnProcess))` only appends if not present; no test verifies a grant without SpawnProcess in its effects still receives it in the lease under a `SpawnCapability` action.

### Performance/Concurrency
- **Medium** Policy parsed from TOML on every `lease_terms_for` call (`crates/ironclaw_reborn_composition/src/runtime/approval.rs:36-40`, confidence 85) — `toml::from_str` + validation runs per approval; the embedded TOML is compile-time constant and the result should be cached at construction time or via `OnceLock`.

### Maintainability
- **Medium** `local_dev_capability_policy()` has no shared cache — repeated static parse from three independent call sites (`crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:1762-1773`, confidence 80) — natural `OnceLock<LocalDevCapabilityPolicy>` candidate; current design invites more callers without noticing the cumulative overhead.

### Conventions
- **Low** `manifest_path` is a plain `String` instead of a domain newtype (`crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:147-148`, confidence 68) — per `.claude/rules/types.md`, paths crossing module boundaries should use a validated newtype; validation only checks `is_empty()` today.

### Tests
- **Low** Bundled policy parse test does not assert `approval_defaults.spawn_capability.mounts` and `.network` values (`crates/ironclaw_reborn_composition/src/local_dev_capability_policy.rs:1880-1918`, confidence 65) — a change to the TOML defaults would not be caught.

### Security
- **Low** `builtin.shell` ambient+`write_filesystem`+`local_dev_wildcard` grant has no explanatory comment in TOML (`crates/ironclaw_reborn_composition/src/local_dev_capability_policy.toml:49-58`, confidence 60) — the old Rust code had a comment explaining the ambient escape-hatch rationale; it was not ported to the declarative TOML, leaving future reviewers without context. Not a regression.

**Review by zmanian — COMMENTED** [review]
## Summary

Refactor that hardens the bundled local-dev capability policy: adds `#[serde(deny_unknown_fields)]` everywhere, strong-types `String` fields into `CapabilityId` / `PackageId` / validated `ExtensionId`, validates effect lists (non-empty + dedup) at parse time, and consolidates the `lease_approval_for` / `builtin_grants` logic into the policy module so `approval.rs` and `local_dev.rs` no longer duplicate the action-to-constraints mapping. Behavior parity with the prior code-driven helpers looks intact.

## Correctness

- Visible-capability surface (`builtin_grants`), skill-management mount selection, and dispatch lease terms all flow through the same parsed policy — no newly-permitted capability vs. the prior `local_dev_grant_constraints` / `local_dev_spawn_capability_constraints` helpers.
- `SpawnCapability` still gets `EffectKind::SpawnProcess` injected via the `required_effect` arg in `constraint_terms`, matching the prior explicit push in `approval.rs`. Spawn fallback to `approval_defaults.spawn_capability` for capabilities not in the grants list is preserved.
- `LocalDevApprovalPolicyAction::from_host_action` returning `None` for non-Dispatch/SpawnCapability actions is mapped to `UnsupportedAction` in `approval.rs` — fail-closed against new action variants.

## Security

- Fail-closed posture is good: `deny_unknown_fields` blocks TOML drift from silently widening grants; `validate_policy` rejects empty grants, duplicate capabilities, empty/duplicate effect lists, empty manifest path, and invalid provider extension id at `include_str!` parse time (effectively at compile/startup, not at request time). Missing-grant on dispatch maps to `LeaseTermsUnavailable` — request is rejected, no implicit grant.
- Local-dev scoping is preserved: `local_dev_capability_policy` module is `mod` (private), only consumed from `build_local_dev_reborn_services` (via `local_dev_first_party_trust_policy`) and `LocalDevApprovalLeaseTermsProvider` / `LocalDevLoopCapabilityPortFactory`. No leakage into production composition paths in this diff.
- `local_dev_wildcard_network_policy` keeps `deny_private_ip_ranges = true`, so the wildcard for shell/http still blocks loopback, link-local, cloud metadata, and private ranges through the shared enforcer.
- No new `info!` / `warn!` in background paths.

## Recommendations

- (Minor) The spawn-capability fallback path inside `lease_approval_for` swallows non-`MissingGrant` errors by mapping them to `LeaseTermsUnavailable` upstream. That's correct, but consider `debug!`-logging the discriminant when the fallback fires so policy drift (e.g. a typo on a capability id that's expected to have an explicit grant) is observable rather than silently routed to defaults.
- (Maintainability) The TOML is the source of truth for what local-dev capabilities exist, but the corresponding `*_CAPABILITY_ID` constants live in `ironclaw_host_runtime`. The test `local_dev_builtin_surface_grants_capability_classes` already asserts the major ones are present — worth keeping that test broad as the registry evolves to catch drift between the TOML and the registry's actual capability id constants.

## Verdict

COMMENT — no blocking issues. The tightening is a strict improvement over the prior code-driven assembly; deny-by-default invariants are preserved and the consolidation removes the previous duplication of effect-injection logic across modules.

**Comment by serrrfirat** [comment]
Addressed the review comments in `485fda421ec7183aefa76dacbd7f4205cfbb0fc4` and the remaining Zmanian observability recommendation in `840eb958bcf7fd1ef78cac94eb0e41173274247c`.

## Fixed Review Feedback

| Reviewer | Direct quote | Fix |
|---|---|---|
| gemini-code-assist[bot] | “parse and validate it exactly once using `std::sync::OnceLock`” | Added a `OnceLock` cache for the bundled local-dev capability policy. |
| henrypark133 | “Policy parsed from TOML on every `lease_terms_for` call.” | `LocalDevApprovalLeaseTermsProvider` now receives the parsed policy at construction time. |
| henrypark133 | “Policy re-parsed on every capability port creation” | `LocalDevLoopCapabilityPortFactory` now stores the parsed policy and skill mounts instead of reparsing per turn. |
| henrypark133 | “Policy parse error is silently swallowed” | Added `tracing::error!` when policy validation fails during runtime build and when approval lease terms cannot be derived. |
| henrypark133 | “manifest_path field is a plain String” | Added validation that the provider manifest path is absolute, beyond the previous non-empty check. |
| henrypark133 | “missing test for SpawnProcess injection” | Added tests that `SpawnCapability` injects `SpawnProcess` for grants without it and does not duplicate it when already declared. |
| henrypark133 | “does not validate approval_defaults.spawn_capability mounts/network fields” | Added assertions for `Workspace` mounts and `Default` network on spawn approval defaults. |
| henrypark133 | “builtin.shell ambient+`write_filesystem`+`local_dev_wildcard` grant has no explanatory comment” | Restored the local-dev shell grant rationale directly in `local_dev_capability_policy.toml`. |
| zmanian | “consider `debug!`-logging the discriminant when the fallback fires” | Added a `tracing::debug!` log with the capability id when spawn-capability approval falls back to default lease terms. |
| zmanian | “worth keeping that test broad as the registry evolves” | Kept the existing broad built-in surface drift test intact while adding the narrower spawn/default policy tests above. |

## Validation

- `cargo fmt -p ironclaw_reborn_composition`
- `cargo test -p ironclaw_reborn_composition`
- `cargo test -p ironclaw_reborn_composition local_dev_capability_policy --lib`
- `cargo clippy -p ironclaw_reborn_composition --all-targets --all-features`
- `git diff --check`

GitHub checks were green for `classify` and `scope` after the first push; they are expected to rerun for the latest follow-up commit.

**Comment by serrrfirat** [comment]
Conflict follow-up landed in `bd5b643909e8b0fa0f9c42db0545eabe15acd22c`.

Merged `origin/reborn-integration` into this PR to clear the GitHub conflict while preserving the base branch extension lifecycle, durable display preview, and capability-host registry changes. The local-dev capability surface remains driven from `local_dev_capability_policy.toml`, with policy entries added for the new base-branch capabilities (`builtin.http.save`, extension lifecycle tools, and skill install network access).

## Validation

- `cargo test -p ironclaw_reborn_composition local_dev_capability_policy --lib`
- `cargo test -p ironclaw_reborn_composition`
- `cargo test -p ironclaw_host_runtime host_runtime_spawn_process_sandbox --test host_runtime_services_contract`
- `cargo test -p ironclaw_loop_support process_sandbox_capability --lib`
- `git diff --check`

`cargo clippy -p ironclaw_reborn_composition -p ironclaw_host_runtime -p ironclaw_loop_support --all-targets --all-features` was attempted locally but the machine ran out of disk before rustc could write metadata; GitHub checks are now running on the pushed merge commit.
